### PR TITLE
webdataset bugfixes, optional wandb logging for VAE training, VQGAN discriminator warmup + spectral norm

### PR DESCRIPTION
**Pull request by GallagherCommaJack**
Had a bunch of pieces that I wanted to change for my personal VQGAN training run, not sure if you want all of them but I'm pretty sure they're all positive.

For VQGAN, hinge loss means that there's not much reason not to train the discriminator in the early steps, before we start to backprop it through the generator, especially since we're going to do the forward passes anyway with the way the current training loop is set up. I also added spectral normalization to the discriminator, to keep the discriminator-induced gradients from exploding too hard.

The other changes are more strictly positive - there were some bugs in webdataset loading leading to input being batches of batches instead of just, well, batches. These changes might need some additional patching to support conditional generators / discriminators, though since the original code didn't work at all I still see this as a strict improvement.

I also added optional wandb logging, which is conditionally imported only if the --wandb argument is passed to train_vae.py, which both logs gradients of the VQGAN and logs reconstructed images in a table instead of in separate grids.

### CODE REVIEW

**Comment by tgisaturday** [comment]
@GallagherCommaJack Thank you for your hard work! I'm currently working on EMA VQ and logging issues with large pods. I didn't have chance to go over webdatasets and wandb and your PR helped me a lot.

**Comment by tgisaturday** [comment]
@GallagherCommaJack I've merged your pr just now. I'll let you know if I face any issue.
